src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -842,44 +860,53 @@ private SqlCommand BuildGetChangesCommand(SqlConnection connection, SqlTransacti
-            // * Not Null LeaseExpirationTime AND
-            // * LeaseExpirationTime > Current Time
-            string getLeaseLockedrowCountQuery = $@"
+            // Lease locked:
+            //  * Have attempts remaining (Attempt count < Max attempts)
+            //  * NOT NULL LeaseExpirationTime
+            //  * LeaseExpirationTime > Current Time
+            // Max Attempts reached:
+            //  * NULL LeaseExpirationTime OR LeaseExpirationTime <= Current Time
+            //  * No attempts remaining (Attempt count = Max attempts)
+            string getLeaseLockedOrMaxAttemptRowCountQuery = $@"
                 {AppLockStatements}
 
                 DECLARE @last_sync_version bigint;
                 SELECT @last_sync_version = LastSyncVersion
                 FROM {GlobalStateTableName}
                 WHERE UserFunctionID = '{this._userFunctionId}' AND UserTableID = {this._userTableId};
-
-                SELECT COUNT(*)
+                DECLARE @lease_locked_count int;
+                DECLARE @max_attempts_count int;
+                SELECT 
+                    @lease_locked_count = COUNT(CASE WHEN l.{LeasesTableAttemptCountColumnName} < {MaxChangeProcessAttemptCount} AND l.{LeasesTableLeaseExpirationTimeColumnName} IS NOT NULL AND l.{LeasesTableLeaseExpirationTimeColumnName} > SYSDATETIME() THEN 1 ELSE NULL END),
+                    @max_attempts_count = COUNT(CASE WHEN (l.{LeasesTableLeaseExpirationTimeColumnName} IS NULL OR l.{LeasesTableLeaseExpirationTimeColumnName} <= SYSDATETIME()) AND l.{LeasesTableAttemptCountColumnName} = {MaxChangeProcessAttemptCount} THEN 1 ELSE NULL END)
                 FROM CHANGETABLE(CHANGES {this._userTable.BracketQuotedFullName}, @last_sync_version) AS c
-                LEFT OUTER JOIN {this._bracketedLeasesTableName} AS l ON {leasesTableJoinCondition}
-                WHERE l.{LeasesTableLeaseExpirationTimeColumnName} IS NOT NULL AND l.{LeasesTableLeaseExpirationTimeColumnName} > SYSDATETIME()";
+                LEFT OUTER JOIN {this._bracketedLeasesTableName} AS l ON {leasesTableJoinCondition};
+                IF @lease_locked_count > 0 OR @max_attempts_count > 0
+                BEGIN
+                    SELECT '(' + CAST(@lease_locked_count AS NVARCHAR) + ' found with lease locks and ' + CAST(@max_attempts_count AS NVARCHAR) + ' ignored because they''ve reached the max attempt limit)';
+                END";
             try
             {
-                using (var getLeaseLockedRowCountCommand = new SqlCommand(getLeaseLockedrowCountQuery, connection, transaction))
+                using (var getLeaseLockedOrMaxAttemptsRowCountCommand = new SqlCommand(getLeaseLockedOrMaxAttemptRowCountQuery, connection, transaction))
                 {
-                    var commandSw = Stopwatch.StartNew();
-                    leaseLockedRowsCount = (int)await getLeaseLockedRowCountCommand.ExecuteScalarAsyncWithLogging(this._logger, CancellationToken.None);
```

Comment:
Note I fixed an issue here with not passing in the cancellation token. We should be doing that for all queries we execute so that if the function execution is cancelled we'll properly cancel the queries we're running right away.